File: 2019/day24/python/planet.py
import collections
import copy
import time
import math

class Planet:
  """ Planet class that scans the bugs """
  def __init__(self,file, recursive = False):
    """ Intialize Planet class """
    self.grid = self.read_map(file)
    self.h, self.w = len(self.grid[0]), len(self.grid[0][0])
    self.recursive = recursive
    self.neighbors = self.get_neighbors()
    self.map_config = dict()
    

  def get_neighbors(self):
    """ Go through every grid position and compute its neighbors """

    # Dictionary to keep track of neighbors
    neighbor_dict = collections.defaultdict(list)

    # Search directions
    directions = [[-1,0], [1,0], [0,1], [0,-1]]

    # Middle of the grid
    mid_y, mid_x = self.h // 2, self.w // 2
    
    for row in range(self.h):
      for col in range(self.w):
        for dx,dy in directions:
          row_, col_ = row + dy, col + dx

          # By default set the neighbors in this direction to empty
          neighbors = []

          # Outer neighbors above
          if row_ < 0 and self.recursive:
            neighbors = [[-1, mid_y-1, mid_x]]
          # Outer neighbors below
          elif row_ >= self.h and self.recursive:
            neighbors = [[-1, mid_y+1, mid_x]]
          # Outer neighbors to the left
          elif col_ < 0 and self.recursive:
            neighbors = [[-1, mid_y, mid_x-1]]
          # Outer neighbors to the right
          elif col_ >= self.w and self.recursive:
            neighbors = [[-1, mid_y, mid_x+1]]
          # Handle center of maze when it is recursive
          elif col_ == self.w // 2 and row_ == self.h // 2 and self.recursive:
            # Left inside
            if dx == 1: neighbors = [ [1, y, 0] for y in range(self.h)]
            # Right inside
            if dx == -1: neighbors = [ [1, y, self.w-1] for y in range(self.h)]
            # Top inside
            if dy == 1: neighbors = [[1, 0, x] for x in range(self.w)]
            # Bottom inside
            if dy == -1: neighbors = [[1, self.h-1, x] for x in range(self.w)]
          # Otherwise, we are inside the boundaries  
          elif col_ >= 0 and col_ < self.w and row_ >= 0 and row_ < self.h:
            neighbors = [[0, row_, col_]]

          # Go through the list of found neighbors and append them:
          # [level_difference, y_pos, x_pos]
          for neighbor in neighbors: neighbor_dict[row, col].append(neighbor)

    return neighbor_dict

  def update(self, n_updates = 1):
    """ Apply the bug update step """

    # If grid is recursive, figure out how many levels are needed
    if self.recursive:
      # Min and max levels
      min_level = int(math.floor(-n_updates / 2))
      max_level =  int(math.ceil(n_updates / 2) + 1)
      
      # Expand the dictionary to include levels : [min_level, max_level]
      [self.grid[level] for level in range(min_level,max_level)]

    
    # Run the number of updates
    for update in range(n_updates):

      # Temporary grid used in between updates
      temp_level = collections.defaultdict(lambda: self.empty_grids)
      
      # We do not need to go through the recursion for all levels
      depth = update // 2
      for level in range(-depth - 1, depth + 2):
      

        # Temporary grid for a given level
        temp_grid = []

        for row in range(self.h):
          new_row = ""
          for col in range(self.w):
            # See how many bugs are as neighbors
            neighbor_str = ""
            for level_dir, y, x in self.neighbors[row, col]:
              neighbor_str += self.grid[level+level_dir][y][x]
            n_neighbors = neighbor_str.count('#')
            
            # Update bug tile
            if self.grid[level][row][col] == '#':
              if n_neighbors == 1: new_row += '#'
              else:                new_row += '.'
            # Update empty tile
            else:
              if n_neighbors == 1 or n_neighbors == 2: new_row += '#'
              else:                                    new_row += '.'

          # Assign the row to the grid
          temp_grid.append(new_row)

        # Assign the grid to the level
        temp_level[level] = temp_grid

      # Update the grid for next update
      self.grid = temp_level

      # Check this for part one
      if not self.recursive:
        # Check if the key has been listed before
        if "".join(self.grid[0]) in self.map_config:
          return True
        # Otherwise keep going with iterations
        else:
          self.map_config["".join(self.grid[0])] = 1
          return False

  def part_one(self):
    """ Solve part one by updating the grid until finding same configuration """
    
    while True:
      # If configuration found
      if self.update():
        self.print_map()
        
        # Flatten the string
        strng = "".join([row for row in self.grid[0]])
        # Compute formula
        return sum([2 ** i if val == '#' else 0 for i,val in enumerate(strng)])
      
      
  def part_two(self):
    """ Solve part two by updating the grid 200 times and counts the bugs """
    
    # Update
    self.update(200)

    # Count bugs
    bugs = 0
    for level in self.grid:
      for row in self.grid[level]:
        bugs +=row.count('#')
      # Remove center of the maze as we can't count it
      if self.grid[level][self.h // 2][ self.w // 2] == '#':
        bugs -= 1

    return bugs

  def read_map(self, fileName):
    """ Read the map """
    self.empty_grids = ['.'*5 for x in range(5)]
    grid = collections.defaultdict(lambda:self.empty_grids)
    
    with open(fileName) as f:
      grid[0] = f.read().splitlines()

    return grid

  def print_map(self):
    """ Print the map """
    for row in range(1,len(self.grid[0])-1):
      print("".join(self.grid[0][row][1:-1]))
    

File: 2019/day24/cpp/src/planet.cpp
#include "planet.hpp"

#include <math.h>

#include <algorithm>

#include "utils.hpp"

Planet::Planet(std::string file_name, bool recursive)
    : m_empty_grid(),
      m_grid(),
      m_h(),
      m_w(),
      m_recursive(recursive),
      m_neighbors_map(),
      m_map_config() {
  // Initialize empty grid and read the map (input)
  m_empty_grid = std::vector<std::string>(5, {"....."});
  m_grid = read_map(file_name);

  // Get width and height of grid
  m_h = m_grid[0].data.size();
  m_w = m_grid[0].data[0].size();

  compute_neighbors();
}

std::map<int, grid_struct> Planet::read_map(std::string file_name) {
  auto input = utils::read_file(file_name);
  std::map<int, grid_struct> grid;
  grid[0] = grid_struct(input);

  return grid;
}

// Make a mapping of how neighbors are connected (look at README)
void Planet::compute_neighbors() {
  std::vector<std::vector<int>> directions({{-1, 0}, {1, 0}, {0, 1}, {0, -1}});

  // Mid point treated differently depending on recursive setting
  int mid_y = m_h / 2;
  int mid_x = m_w / 2;

  for (int row = 0; row < m_h; row++) {
    for (int col = 0; col < m_w; col++) {
      std::vector<std::vector<int>> neighbors;
      for (auto &d : directions) {
        int dx = d[0], dy = d[1];
        auto row_ = row + dy, col_ = col + dx;

        if (row_ < 0 && m_recursive) {
          neighbors.push_back({-1, mid_y - 1, mid_x});
        } else if (row_ >= m_h && m_recursive) {
          neighbors.push_back({-1, mid_y + 1, mid_x});
        } else if (col_ < 0 && m_recursive) {
          neighbors.push_back({-1, mid_y, mid_x - 1});
        } else if (col_ >= m_h && m_recursive) {
          neighbors.push_back({-1, mid_y, mid_x + 1});
        } else if (col_ == mid_x && row_ == mid_y && m_recursive) {
          if (dx == 1) {
            for (int y = 0; y < m_h; y++) {
              neighbors.push_back({1, y, 0});
            }
          } else if (dx == -1) {
            for (int y = 0; y < m_h; y++) {
              neighbors.push_back({1, y, m_w - 1});
            }
          } else if (dy == 1) {
            for (int x = 0; x < m_w; x++) {
              neighbors.push_back({1, 0, x});
            }
          } else if (dy == -1) {
            for (int x = 0; x < m_w; x++) {
              neighbors.push_back({1, m_h - 1, x});
            }
          }
        } else if (col_ >= 0 && col_ < m_w && row_ >= 0 && row_ < m_h) {
          neighbors.push_back({0, row_, col_});
        }
      }
      m_neighbors_map[std::make_pair(row, col)] = neighbors;
    }
  }
}

// Perform the updates of the planet
bool Planet::update(int n_updates) {
  // Initialize recursive levels
  int min_level, max_level;
  if (m_recursive) {
    min_level = int(floor(-n_updates / 2));
    max_level = int(ceil(n_updates / 2) + 1);
    for (int level = min_level; level < max_level; level++) {
      m_grid[level];
    }
  }

  // Loop through updates
  for (int u = 0; u < n_updates; u++) {
    std::map<int, grid_struct> temp_level;

    int depth = u / 2;

    for (int level = -depth - 1; level < depth + 2; level++) {
      std::vector<std::string> temp_grid;

      for (int row = 0; row < m_h; row++) {
        std::string new_row("");
        for (int col = 0; col < m_w; col++) {
          // Add all neighbors in a string
          std::string nbors = "";
          for (auto &level_y_x : m_neighbors_map[std::make_pair(row, col)]) {
            auto level_dir = level_y_x[0];
            auto y = level_y_x[1];
            auto x = level_y_x[2];
            nbors += m_grid[level + level_dir].data[y][x];
          }

          // See how many neighbors are present ( number of '#') and update
          size_t n_neighbors = std::count(nbors.begin(), nbors.end(), '#');
          if (m_grid[level].data[row][col] == '#') {
            if (n_neighbors == 1) {
              new_row += '#';
            } else {
              new_row += '.';
            }
          } else {
            if (n_neighbors == 1 || n_neighbors == 2) {
              new_row += '#';
            } else {
              new_row += '.';
            }
          }
        }
        // Update the temporary grid
        temp_grid.push_back(new_row);
      }
      // Update the temporary level
      temp_level[level] = temp_grid;
    }
    // Update grid after the update step
    m_grid = temp_level;

    // For part one, check if level=0 is repeated
    if (!m_recursive) {
      std::string str;
      for (auto &x : m_grid[0].data) {
        str += x;
      }
      if (m_map_config.count(str)) {
        return true;
      } else {
        m_map_config.insert(str);
      }
    }
  }
  return false;
}

// Solve part one: update the system until it is repeated
int Planet::part_one() {
  int iter(0);
  while (true) {
    // If sequence is repeated compute the metric
    if (update()) {
      std::string strng;
      for (auto &x : m_grid[0].data) {
        strng += x;
      }
      int sm(0);
      for (size_t i = 0; i < strng.size(); i++) {
        sm += (strng[i] == '#') ? pow(2, i) : 0;
      }
      return sm;
    }
    iter++;
  }
}

// Solve part two: Update the system n_steps times and count how many #
// across all layers
int Planet::part_two(int n_steps) {
  update(n_steps);
  size_t bugs = 0;
  for (auto &level : m_grid) {
    for (auto &row : level.second.data) {
      bugs += std::count(row.begin(), row.end(), '#');
    }
    if (level.second.data[m_h / 2][m_w / 2] == '#') {
      bugs--;
    }
  }
  return bugs;
}
